File: scripts/ai-workflow/phases/testing.py
"""Phase 5: テスト実行フェーズ

Phase 4で実装したテストコードを実行し、結果を記録する。
テスト失敗時はPhase 4に戻って修正が必要。
"""
import re
from pathlib import Path
from typing import Dict, Any, List
from phases.base.abstract_phase import AbstractPhase


class TestingPhase(AbstractPhase):
    """テスト実行フェーズ"""

    def __init__(self, *args, **kwargs):
        """初期化"""
        super().__init__(
            phase_name='testing',
            *args,
            **kwargs
        )

    def execute(self) -> Dict[str, Any]:
        """
        テスト実行フェーズを実行

        Returns:
            Dict[str, Any]: 実行結果
                - success: bool
                - output: str - test-result.mdのパス
                - error: Optional[str]
        """
        try:
            # Issue情報を取得
            issue_number = int(self.metadata.data['issue_number'])

            # 必要なファイルのパスを定義
            test_implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '05_test_implementation' / 'output' / 'test-implementation.md'
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'

            if not test_implementation_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'テスト実装ログが見つかりません: {test_implementation_file}'
                }

            if not implementation_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'実装ログが見つかりません: {implementation_file}'
                }

            if not test_scenario_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'テストシナリオが見つかりません: {test_scenario_file}'
                }

            # Planning Phase成果物のパス取得
            planning_path_str = self._get_planning_document_path(issue_number)

            # 実行プロンプトを読み込み
            execute_prompt_template = self.load_prompt('execute')

            # working_dirからの相対パスを使用
            rel_path_test_implementation = test_implementation_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            execute_prompt = execute_prompt_template.replace(
                '{planning_document_path}',
                planning_path_str
            ).replace(
                '{test_implementation_document_path}',
                f'@{rel_path_test_implementation}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # test-result.mdのパスを取得
            output_file = self.output_dir / 'test-result.md'

            # 既存ファイルの最終更新時刻を記録（上書き確認用）
            old_mtime = output_file.stat().st_mtime if output_file.exists() else None

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=execute_prompt,
                max_turns=30,
                log_prefix='execute'
            )

            # test-result.mdが存在するか確認
            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': f'test-result.mdが生成されませんでした: {output_file}'
                }

            # ファイルが更新されたか確認（タイムスタンプチェック）
            new_mtime = output_file.stat().st_mtime
            if old_mtime is not None and new_mtime == old_mtime:
                return {
                    'success': False,
                    'output': None,
                    'error': f'test-result.mdが更新されませんでした（古いファイルのまま）: {output_file}'
                }

            # GitHub Issueに成果物を投稿
            try:
                output_content = output_file.read_text(encoding='utf-8')
                self.post_output(
                    output_content=output_content,
                    title="テスト結果"
                )
            except Exception as e:
                print(f"[WARNING] 成果物のGitHub投稿に失敗しました: {e}")

            # ステータス更新: BasePhase.run()で実行されるため不要
            # self.metadata.update_phase_status('testing', 'completed', str(output_file))
            # self.post_progress('completed', f'テストが完了しました: {output_file.name}')

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            # ステータス更新: 失敗
            self.metadata.update_phase_status('testing', 'failed')
            # BasePhase.run()で実行されるため不要
            # self.post_progress('failed', f'テストが失敗しました: {str(e)}')

            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

    def review(self) -> Dict[str, Any]:
        """
        テスト結果をレビュー

        Returns:
            Dict[str, Any]: レビュー結果
                - result: str - PASS/PASS_WITH_SUGGESTIONS/FAIL
                - feedback: str
                - suggestions: List[str]
        """
        try:
            # test-result.mdを読み込み
            test_result_file = self.output_dir / 'test-result.md'

            if not test_result_file.exists():
                return {
                    'result': 'FAIL',
                    'feedback': 'test-result.mdが存在しません。',
                    'suggestions': ['execute()を実行してtest-result.mdを生成してください。']
                }

            # 実装ログとテストシナリオのパス
            issue_number = int(self.metadata.data['issue_number'])
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'

            # レビュープロンプトを読み込み
            review_prompt_template = self.load_prompt('review')

            # working_dirからの相対パスを使用
            rel_path_test_result = test_result_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            review_prompt = review_prompt_template.replace(
                '{test_result_document_path}',
                f'@{rel_path_test_result}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            )

            # Claude Agent SDKでレビューを実行
            messages = self.execute_with_claude(
                prompt=review_prompt,
                max_turns=30,
                log_prefix='review'
            )

            # レビュー結果をパース
            review_result = self._parse_review_result(messages)

            # レビュー結果をファイルに保存
            review_file = self.review_dir / 'result.md'
            review_file.write_text(review_result['feedback'], encoding='utf-8')
            print(f"[INFO] レビュー結果を保存: {review_file}")

            # GitHub Issueにレビュー結果を投稿: BasePhase.run()で実行されるため不要
            # self.post_review(
            #     result=review_result['result'],
            #     feedback=review_result['feedback'],
            #     suggestions=review_result.get('suggestions')
            # )

            return review_result

        except Exception as e:
            return {
                'result': 'FAIL',
                'feedback': f'レビュー中にエラーが発生しました: {str(e)}',
                'suggestions': []
            }

    def revise(self, review_feedback: str) -> Dict[str, Any]:
        """
        レビュー結果を元にテストを再実行または実装を修正

        Args:
            review_feedback: レビュー結果のフィードバック

        Returns:
            Dict[str, Any]: 修正結果
                - success: bool
                - output: str - test-result.mdのパス
                - error: Optional[str]
        """
        try:
            # 元のテスト結果のパス
            test_result_file = self.output_dir / 'test-result.md'

            if not test_result_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': 'test-result.mdが存在しません。'
                }

            # 必要なファイルのパスを定義
            issue_number = int(self.metadata.data['issue_number'])
            test_implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '05_test_implementation' / 'output' / 'test-implementation.md'
            implementation_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '04_implementation' / 'output' / 'implementation.md'
            test_scenario_file = self.metadata.workflow_dir.parent / f'issue-{issue_number}' / '03_test_scenario' / 'output' / 'test-scenario.md'

            # 修正プロンプトを読み込み
            revise_prompt_template = self.load_prompt('revise')

            # working_dirからの相対パスを使用
            rel_path_test_result = test_result_file.relative_to(self.claude.working_dir)
            rel_path_test_implementation = test_implementation_file.relative_to(self.claude.working_dir)
            rel_path_implementation = implementation_file.relative_to(self.claude.working_dir)
            rel_path_test_scenario = test_scenario_file.relative_to(self.claude.working_dir)

            # プロンプトに情報を埋め込み
            revise_prompt = revise_prompt_template.replace(
                '{test_result_document_path}',
                f'@{rel_path_test_result}'
            ).replace(
                '{review_feedback}',
                review_feedback
            ).replace(
                '{test_implementation_document_path}',
                f'@{rel_path_test_implementation}'
            ).replace(
                '{implementation_document_path}',
                f'@{rel_path_implementation}'
            ).replace(
                '{test_scenario_document_path}',
                f'@{rel_path_test_scenario}'
            ).replace(
                '{issue_number}',
                str(issue_number)
            )

            # 既存ファイルの最終更新時刻を記録（上書き確認用）
            old_mtime = test_result_file.stat().st_mtime if test_result_file.exists() else None

            # Claude Agent SDKでタスクを実行
            messages = self.execute_with_claude(
                prompt=revise_prompt,
                max_turns=30,
                log_prefix='revise'
            )

            # test-result.mdが存在するか確認
            output_file = self.output_dir / 'test-result.md'

            if not output_file.exists():
                return {
                    'success': False,
                    'output': None,
                    'error': '修正されたtest-result.mdが生成されませんでした。'
                }

            # ファイルが更新されたか確認（タイムスタンプチェック）
            new_mtime = output_file.stat().st_mtime
            if old_mtime is not None and new_mtime == old_mtime:
                return {
                    'success': False,
                    'output': None,
                    'error': f'test-result.mdが更新されませんでした（古いファイルのまま）: {output_file}'
                }

            return {
                'success': True,
                'output': str(output_file),
                'error': None
            }

        except Exception as e:
            return {
                'success': False,
                'output': None,
                'error': str(e)
            }

File: scripts/ai-workflow-v2/src/phases/testing.ts
import fs from 'fs-extra';
import path from 'node:path';
import { BasePhase, type PhaseInitializationParams } from './base-phase.js';
import { PhaseExecutionResult } from '../types.js';

export class TestingPhase extends BasePhase {
  constructor(params: PhaseInitializationParams) {
    super({ ...params, phaseName: 'testing' });
  }

  protected async execute(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const planningReference = this.getPlanningDocumentReference(issueNumber);

    // オプショナルコンテキストを構築（Issue #398）
    const testImplementationContext = this.buildOptionalContext(
      'test_implementation',
      'test-implementation.md',
      'テストコード実装ログは利用できません。実装コードを直接確認してテストを実行してください。',
      issueNumber,
    );

    const implementationContext = this.buildOptionalContext(
      'implementation',
      'implementation.md',
      '実装ログは利用できません。リポジトリの実装コードを直接確認してください。',
      issueNumber,
    );

    const scenarioContext = this.buildOptionalContext(
      'test_scenario',
      'test-scenario.md',
      'テストシナリオは利用できません。実装内容に基づいて適切なテストを実施してください。',
      issueNumber,
    );

    const testResultFile = path.join(this.outputDir, 'test-result.md');
    const oldMtime = fs.existsSync(testResultFile) ? fs.statSync(testResultFile).mtimeMs : null;

    const executePrompt = this.loadPrompt('execute')
      .replace('{planning_document_path}', planningReference)
      .replace('{test_implementation_context}', testImplementationContext)
      .replace('{implementation_context}', implementationContext)
      .replace('{test_scenario_context}', scenarioContext)
      .replace('{issue_number}', String(issueNumber));

    await this.executeWithAgent(executePrompt, { maxTurns: 50 });

    if (!fs.existsSync(testResultFile)) {
      return {
        success: false,
        error: `test-result.md が見つかりません: ${testResultFile}`,
      };
    }

    const newMtime = fs.statSync(testResultFile).mtimeMs;
    if (oldMtime !== null && newMtime === oldMtime) {
      return {
        success: false,
        error: 'test-result.md が更新されていません。出力内容を確認してください。',
      };
    }

    return {
      success: true,
      output: testResultFile,
    };
  }

  protected async review(): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const testResultFile = path.join(this.outputDir, 'test-result.md');

    if (!fs.existsSync(testResultFile)) {
      return {
        success: false,
        error: 'test-result.md が存在しません。execute() を先に実行してください。',
      };
    }

    const testImplementationFile = this.getPhaseOutputFile(
      'test_implementation',
      'test-implementation.md',
      issueNumber,
    );
    const implementationFile = this.getPhaseOutputFile('implementation', 'implementation.md', issueNumber);
    const scenarioFile = this.getPhaseOutputFile('test_scenario', 'test-scenario.md', issueNumber);

    const testResultRef = this.getAgentFileReference(testResultFile);
    if (!testResultRef) {
      return {
        success: false,
        error: 'Agent が test-result.md を参照できません。',
      };
    }

    // test_implementation, implementation, scenario はオプショナル（Issue #405）
    let testImplementationRef: string;
    if (testImplementationFile) {
      const ref = this.getAgentFileReference(testImplementationFile);
      testImplementationRef = ref ?? 'テストコード実装ログは利用できません。テスト結果から実装を推測してレビューしてください。';
    } else {
      testImplementationRef = 'テストコード実装ログは利用できません。テスト結果から実装を推測してレビューしてください。';
    }

    let implementationRef: string;
    if (implementationFile) {
      const ref = this.getAgentFileReference(implementationFile);
      implementationRef = ref ?? '実装ログは利用できません。テスト結果から実装を推測してレビューしてください。';
    } else {
      implementationRef = '実装ログは利用できません。テスト結果から実装を推測してレビューしてください。';
    }

    let scenarioRef: string;
    if (scenarioFile) {
      const ref = this.getAgentFileReference(scenarioFile);
      scenarioRef = ref ?? 'テストシナリオは利用できません。テスト結果から適切なテスト観点でレビューしてください。';
    } else {
      scenarioRef = 'テストシナリオは利用できません。テスト結果から適切なテスト観点でレビューしてください。';
    }

    const reviewPrompt = this.loadPrompt('review')
      .replace('{test_result_document_path}', testResultRef)
      .replace('{test_implementation_document_path}', testImplementationRef)
      .replace('{implementation_document_path}', implementationRef)
      .replace('{test_scenario_document_path}', scenarioRef);

    const messages = await this.executeWithAgent(reviewPrompt, { maxTurns: 50, logDir: this.reviewDir });
    const reviewResult = await this.contentParser.parseReviewResult(messages);

    const reviewFile = path.join(this.reviewDir, 'result.md');
    fs.writeFileSync(reviewFile, reviewResult.feedback, 'utf-8');

    await this.github.postReviewResult(
      issueNumber,
      this.phaseName,
      reviewResult.result,
      reviewResult.feedback,
      reviewResult.suggestions,
    );

    return {
      success: reviewResult.result !== 'FAIL',
      output: reviewResult.result,
      error: reviewResult.result === 'FAIL' ? reviewResult.feedback : undefined,
    };
  }

  public async revise(reviewFeedback: string): Promise<PhaseExecutionResult> {
    const issueNumber = parseInt(this.metadata.data.issue_number, 10);
    const testResultFile = path.join(this.outputDir, 'test-result.md');

    if (!fs.existsSync(testResultFile)) {
      return {
        success: false,
        error: 'test-result.md が存在しません。execute() を先に実行してください。',
      };
    }

    const testImplementationFile = this.getPhaseOutputFile(
      'test_implementation',
      'test-implementation.md',
      issueNumber,
    );
    const implementationFile = this.getPhaseOutputFile('implementation', 'implementation.md', issueNumber);
    const scenarioFile = this.getPhaseOutputFile('test_scenario', 'test-scenario.md', issueNumber);

    const testResultRef = this.getAgentFileReference(testResultFile);
    if (!testResultRef) {
      return {
        success: false,
        error: 'Agent が test-result.md を参照できません。',
      };
    }

    // test_implementation, implementation, scenario はオプショナル（Issue #405）
    let testImplementationRef: string;
    if (testImplementationFile) {
      const ref = this.getAgentFileReference(testImplementationFile);
      testImplementationRef = ref ?? 'テストコード実装ログは利用できません。テスト結果から実装を推測してください。';
    } else {
      testImplementationRef = 'テストコード実装ログは利用できません。テスト結果から実装を推測してください。';
    }

    let implementationRef: string;
    if (implementationFile) {
      const ref = this.getAgentFileReference(implementationFile);
      implementationRef = ref ?? '実装ログは利用できません。テスト結果から実装を推測してください。';
    } else {
      implementationRef = '実装ログは利用できません。テスト結果から実装を推測してください。';
    }

    let scenarioRef: string;
    if (scenarioFile) {
      const ref = this.getAgentFileReference(scenarioFile);
      scenarioRef = ref ?? 'テストシナリオは利用できません。テスト結果から適切なテスト観点で修正してください。';
    } else {
      scenarioRef = 'テストシナリオは利用できません。テスト結果から適切なテスト観点で修正してください。';
    }

    const revisePrompt = this.loadPrompt('revise')
      .replace('{test_result_document_path}', testResultRef)
      .replace('{test_implementation_document_path}', testImplementationRef)
      .replace('{implementation_document_path}', implementationRef)
      .replace('{test_scenario_document_path}', scenarioRef)
      .replace('{review_feedback}', reviewFeedback)
      .replace('{issue_number}', String(issueNumber));

    const oldMtime = fs.existsSync(testResultFile) ? fs.statSync(testResultFile).mtimeMs : null;

    await this.executeWithAgent(revisePrompt, { maxTurns: 50, logDir: this.reviseDir });

    if (!fs.existsSync(testResultFile)) {
      return {
        success: false,
        error: '改訂後の test-result.md を確認できませんでした。',
      };
    }

    const newMtime = fs.statSync(testResultFile).mtimeMs;
    if (oldMtime !== null && newMtime === oldMtime) {
      return {
        success: false,
        error: 'test-result.md が更新されていません。再度実行内容を確認してください。',
      };
    }

    return {
      success: true,
      output: testResultFile,
    };
  }
}
